_posts/2020-08-11-relay-pagingation-containers-demystified.markdown

```diff
@@ -0,0 +1,234 @@
+  for a certain set of variables.
+- `getVariables` is used when actually fetching another page, and its return value is given to the `query`.
+
+I think of `getFragmentVariables` as kind of cache key for lookup in Relay's internal store. Our implementation of
+`getFragmentVariables` above doesn't really do anything interesting, but a connection that accepted `sort` or
+`filter` parameters would need to return those to avoid lookup collisions when the user changed sort and filter
+options.
+
+Now for `getVariables`, which are the variables used for the `query` later on. It really ought to be named
+`getQueryVariables`, I think. But I digress.
+
+Every implementation of `getFragmentVariables` I could find at Artsy was identical, which makes sense because _that
+is the default implementation_. We shouldn't be defining this option at all! As far as I can tell, Artsy started
+with a few pagination containers that supplied this parameter unnecessarily and it got copy-and-pasted throughout
+our codebases.
+
+After revisiting the docs, I noticed other optional parameters that don't need to be defined either. Let's rewrite
+the call to `createPaginationContainer` to only supply the parameters that are required:
+
+```js
+export ArtistFragmentContainer = createPaginationContainer(Artist, {
+  artist: graphql`
+    fragment Artist_artist on Artist @argumentDefinitions(
+      count: { type: "Int", defaultValue: 10 }
+      cursor: { type: "String" } {
+      name
+      id
+      artworksConnection (first: $count, after: $cursor) @connection(key: "Artist_artworks") {
+        edges {
+          node {
+            title # Now fetch all the artwork data
+            id
+          }
+        }
+      }
+    }
+  `
+  }, {
+  getVariables(props, { count, cursor }, fragmentVariables) {
+    return {
+      id: props.artist.id,
+      count,
+      cursor,
+    }
+  },
+  query: graphql` # Here is the query to fetch any specific page
+    query ArtistArtworksQuery(
+      $id: ID!
+      $count: Int!
+      $cursor: String) {
+      artist(id: $id) {
+        ...Artist_artist @arguments(
+          count: $count
+          cursor: $cursor
+        )
+      }
+    }
+  `
+})
+```
+
+This is a lot nicer! By not specifying unnecessary options, we have a smaller surface area to make mistakes in. We
+also have fewer overloaded terms, like "variables", so now it's more obvious that `getVariables` supplies data for
+the `query` below it.
+
+Needless to say that I'll be following up with some pull requests to clean up our use of Relay pagination
+containers. But I wouldn't have discovered this if I hadn't really dug into the docs, which I only did so that I
+could write this blog post. Earlier I said that the solution to a knowledge gap is simple: learn, and then teach. I
+learned a lot about Relay today, and I hope this blog post illustrates the value in the learn-then-teach approach.
```

Comment:
**praise:** This is the kind of high quality content that makes me happy to be an engineer at Artsy. Props @ashfurrow! 🙌

Reply:
+1 good write-up

Reply:
Yup, very nice! 🙌 Wondering if you see any opportunities to improve the upstream docs too?

Reply:
The docs are generally really great, and they do already point out where the examples use the default implementations. Once I ship this, I'll go back and re-read the docs to look for opportunities to improve them 👍